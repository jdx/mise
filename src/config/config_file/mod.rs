use std::collections::{HashMap, HashSet};
use std::ffi::OsStr;
use std::fmt::{Debug, Display};
use std::hash::Hash;
use std::path::{Path, PathBuf};
use std::sync::Mutex;

use eyre::Result;
use once_cell::sync::Lazy;
use serde_derive::Deserialize;
use versions::Versioning;

use tool_versions::ToolVersions;

use crate::cli::args::{ForgeArg, ToolArg};
use crate::config::config_file::mise_toml::MiseToml;
use crate::config::env_directive::EnvDirective;
use crate::config::{AliasMap, Settings};
use crate::errors::Error::UntrustedConfig;
use crate::file::display_path;
use crate::hash::{file_hash_sha256, hash_to_str};
use crate::task::Task;
use crate::toolset::{ToolVersionList, Toolset};
use crate::ui::{prompt, style};
use crate::{dirs, env, file, forge};

pub mod legacy_version;
pub mod mise_toml;
pub mod toml;
pub mod tool_versions;

#[derive(Debug, PartialEq)]
pub enum ConfigFileType {
    MiseToml,
    ToolVersions,
    // LegacyVersion,
}

pub trait ConfigFile: Debug + Send + Sync {
    fn get_path(&self) -> &Path;
    fn min_version(&self) -> &Option<Versioning> {
        &None
    }
    /// gets the project directory for the config
    /// if it's a global/system config, returns None
    /// files like ~/src/foo/.mise/config.toml will return ~/src/foo
    /// and ~/src/foo/.mise.config.toml will return None
    fn project_root(&self) -> Option<&Path> {
        let p = self.get_path();
        if *env::MISE_GLOBAL_CONFIG_FILE == p {
            return None;
        }
        match p.parent() {
            Some(dir) => match dir {
                dir if dir.starts_with(*dirs::CONFIG) => None,
                dir if dir.starts_with(*dirs::SYSTEM) => None,
                dir if dir == *dirs::HOME => None,
                dir => Some(dir),
            },
            None => None,
        }
    }
    fn plugins(&self) -> HashMap<String, String> {
        Default::default()
    }
    fn env_entries(&self) -> Vec<EnvDirective> {
        Default::default()
    }
    fn tasks(&self) -> Vec<&Task> {
        Default::default()
    }
    fn remove_plugin(&mut self, _fa: &ForgeArg) -> Result<()>;
    fn replace_versions(&mut self, fa: &ForgeArg, versions: &[String]) -> Result<()>;
    fn save(&self) -> Result<()>;
    fn dump(&self) -> Result<String>;
    fn to_toolset(&self) -> Result<Toolset>;
    fn aliases(&self) -> AliasMap {
        Default::default()
    }
    fn task_config(&self) -> &TaskConfig {
        static DEFAULT_TASK_CONFIG: Lazy<TaskConfig> = Lazy::new(TaskConfig::default);
        &DEFAULT_TASK_CONFIG
    }
}

impl dyn ConfigFile {
    pub fn add_runtimes(&mut self, tools: &[ToolArg], pin: bool) -> Result<()> {
        // TODO: this has become a complete mess and could probably be greatly simplified
        let mut ts = self.to_toolset()?.to_owned();
        ts.resolve();
        let mut plugins_to_update = HashMap::new();
        for ta in tools {
            if let Some(tv) = &ta.tvr {
                plugins_to_update
                    .entry(ta.forge.clone())
                    .or_insert_with(Vec::new)
                    .push(tv);
            }
        }
        for (fa, versions) in &plugins_to_update {
            let mut tvl = ToolVersionList::new(fa.clone(), ts.source.as_ref().unwrap().clone());
            for tv in versions {
                tvl.requests.push(((*tv).clone(), Default::default()));
            }
            ts.versions.insert(fa.clone(), tvl);
        }
        ts.resolve();
        for (fa, versions) in plugins_to_update {
            let versions = versions
                .into_iter()
                .map(|tvr| {
                    if pin {
                        let plugin = forge::get(&fa);
                        let tv = tvr.resolve(plugin.as_ref(), Default::default(), false)?;
                        Ok(tv.version)
                    } else {
                        Ok(tvr.version())
                    }
                })
                .collect::<Result<Vec<_>>>()?;
            self.replace_versions(&fa, &versions)?;
        }

        Ok(())
    }

    /// this is for `mise local|global TOOL` which will display the version instead of setting it
    /// it's only valid to use a single tool in this case
    /// returns "true" if the tool was displayed which means the CLI should exit
    pub fn display_runtime(&self, runtimes: &[ToolArg]) -> Result<bool> {
        // in this situation we just print the current version in the config file
        if runtimes.len() == 1 && runtimes[0].tvr.is_none() {
            let fa = &runtimes[0].forge;
            let tvl = self
                .to_toolset()?
                .versions
                .get(fa)
                .ok_or_else(|| {
                    eyre!(
                        "no version set for {} in {}",
                        fa.to_string(),
                        display_path(self.get_path())
                    )
                })?
                .requests
                .iter()
                .map(|(tvr, _)| tvr.version())
                .collect::<Vec<_>>();
            miseprintln!("{}", tvl.join(" "));
            return Ok(true);
        }
        // check for something like `mise local node python@latest` which is invalid
        if runtimes.iter().any(|r| r.tvr.is_none()) {
            return Err(eyre!("invalid input, specify a version for each tool. Or just specify one tool to print the current version"));
        }
        Ok(false)
    }
}

fn init(path: &Path) -> Box<dyn ConfigFile> {
    match detect_config_file_type(path) {
        Some(ConfigFileType::MiseToml) => Box::new(MiseToml::init(path)),
        Some(ConfigFileType::ToolVersions) => Box::new(ToolVersions::init(path)),
        _ => panic!("Unknown config file type: {}", path.display()),
    }
}

pub fn parse_or_init(path: &Path) -> Result<Box<dyn ConfigFile>> {
    let cf = match path.exists() {
        true => parse(path)?,
        false => init(path),
    };
    Ok(cf)
}

pub fn parse(path: &Path) -> Result<Box<dyn ConfigFile>> {
    if let Ok(settings) = Settings::try_get() {
        if settings.paranoid {
            trust_check(path)?;
        }
    }
    match detect_config_file_type(path) {
        Some(ConfigFileType::MiseToml) => Ok(Box::new(MiseToml::from_file(path)?)),
        Some(ConfigFileType::ToolVersions) => Ok(Box::new(ToolVersions::from_file(path)?)),
        #[allow(clippy::box_default)]
        _ => Ok(Box::new(MiseToml::default())),
    }
}

pub fn trust_check(path: &Path) -> Result<()> {
    let default_cmd = String::new();
    let args = env::ARGS.read().unwrap();
    let cmd = args.get(1).unwrap_or(&default_cmd).as_str();
    if is_trusted(path) || cmd == "trust" || cfg!(test) {
        return Ok(());
    }
    if cmd != "hook-env" {
        let ans = prompt::confirm(format!(
            "{} {} is not trusted. Trust it?",
            style::eyellow("mise"),
            style::epath(path)
        ))?;
        if ans {
            trust(path)?;
            return Ok(());
        }
    }
    Err(UntrustedConfig())?
}

pub fn is_trusted(path: &Path) -> bool {
    static IS_TRUSTED: Lazy<Mutex<HashSet<PathBuf>>> = Lazy::new(|| Mutex::new(HashSet::new()));
    let mut cached = IS_TRUSTED.lock().unwrap();
    if cached.contains(path) {
        return true;
    }
    let settings = Settings::get();
    for p in settings.trusted_config_paths() {
        if path.starts_with(p) {
            cached.insert(path.to_path_buf());
            return true;
        }
    }
    if settings.paranoid {
        let trusted = trust_file_hash(path).unwrap_or_else(|e| {
            warn!("trust_file_hash: {e}");
            false
        });
        if !trusted {
            return false;
        }
    } else if !(trust_path(path).exists() || ci_info::is_ci() && !cfg!(test)) {
        // the file isn't trusted, and we're not on a CI system where we generally assume we can
        // trust config files
        return false;
    }
    cached.insert(path.to_path_buf());
    true
}

pub fn trust(path: &Path) -> Result<()> {
    let hashed_path = trust_path(path);
    if !hashed_path.exists() {
        file::create_dir_all(hashed_path.parent().unwrap())?;
        file::make_symlink(path, &hashed_path)?;
    }
    let trust_hash_path = hashed_path.with_extension("hash");
    let hash = file_hash_sha256(path)?;
    file::write(trust_hash_path, hash)?;
    Ok(())
}

pub fn untrust(path: &Path) -> Result<()> {
    let hashed_path = trust_path(path);
    if hashed_path.exists() {
        file::remove_file(hashed_path)?;
    }
    Ok(())
}

/// generates a path like ~/.mise/trusted-configs/dir-file-3e8b8c44c3.toml
fn trust_path(path: &Path) -> PathBuf {
    let trust_path = dirs::TRUSTED_CONFIGS.join(hash_to_str(&path));
    if trust_path.exists() {
        return trust_path;
    }
    let trunc_str = |s: &OsStr| {
        let mut s = s.to_str().unwrap().to_string();
        s.truncate(20);
        s
    };
    let parent = path
        .parent()
        .map(|p| p.to_path_buf())
        .unwrap_or_default()
        .file_name()
        .map(trunc_str);
    let filename = path.file_name().map(trunc_str);
    let hash = hash_to_str(&path);

    dirs::TRUSTED_CONFIGS.join(
        [parent, filename, Some(hash)]
            .into_iter()
            .flatten()
            .collect::<Vec<_>>()
            .join("-"),
    )
}

fn trust_file_hash(path: &Path) -> Result<bool> {
    let trust_path = trust_path(path);
    let trust_hash_path = trust_path.with_extension("hash");
    if !trust_hash_path.exists() {
        return Ok(false);
    }
    let hash = file::read_to_string(&trust_hash_path)?;
    let actual = file_hash_sha256(path)?;
    Ok(hash == actual)
}

fn detect_config_file_type(path: &Path) -> Option<ConfigFileType> {
    match path.file_name().unwrap().to_str().unwrap() {
        f if f.ends_with(".toml") => Some(ConfigFileType::MiseToml),
        f if env::MISE_DEFAULT_CONFIG_FILENAME.as_str() == f => Some(ConfigFileType::MiseToml),
        f if env::MISE_DEFAULT_TOOL_VERSIONS_FILENAME.as_str() == f => {
            Some(ConfigFileType::ToolVersions)
        }
        _ => None,
    }
}

impl Display for dyn ConfigFile {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        let toolset = self.to_toolset().unwrap().to_string();
        write!(f, "{}: {toolset}", &display_path(self.get_path()))
    }
}

impl PartialEq for dyn ConfigFile {
    fn eq(&self, other: &Self) -> bool {
        self.get_path() == other.get_path()
    }
}

impl Eq for dyn ConfigFile {}

impl Hash for dyn ConfigFile {
    fn hash<H: std::hash::Hasher>(&self, state: &mut H) {
        self.get_path().hash(state);
    }
}

#[derive(Clone, Debug, Default, Deserialize)]
pub struct TaskConfig {
    pub includes: Option<Vec<PathBuf>>,
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_detect_config_file_type() {
        assert_eq!(
            detect_config_file_type(Path::new("/foo/bar/.test-tool-versions")),
            Some(ConfigFileType::ToolVersions)
        );
        assert_eq!(
            detect_config_file_type(Path::new("/foo/bar/.tool-versions.toml")),
            Some(ConfigFileType::MiseToml)
        );
    }
}
