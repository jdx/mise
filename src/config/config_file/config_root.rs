use std::collections::HashMap;
use std::path::{Path, PathBuf};
use std::sync::{LazyLock as Lazy, Mutex};

use path_absolutize::Absolutize;
use xx::regex;

use crate::config::is_global_config;
use crate::env;

static CONFIG_ROOT_CACHE: Lazy<Mutex<HashMap<PathBuf, PathBuf>>> =
    Lazy::new(|| Mutex::new(HashMap::new()));

pub fn reset() {
    CONFIG_ROOT_CACHE.lock().unwrap().clear();
}

pub fn config_root(path: &Path) -> PathBuf {
    if is_global_config(path) {
        return env::MISE_GLOBAL_CONFIG_ROOT.to_path_buf();
    }
    let path = path
        .absolutize()
        .map(|p| p.to_path_buf())
        .unwrap_or(path.to_path_buf());
    if let Some(cached) = CONFIG_ROOT_CACHE.lock().unwrap().get(&path).cloned() {
        return cached;
    }
    let parts = path
        .components()
        .map(|c| c.as_os_str().to_string_lossy().to_string())
        .collect::<Vec<_>>();
    const EMPTY: &str = "";
    let filename = parts.last().map(|p| p.as_str()).unwrap_or(EMPTY);
    let parent = parts
        .iter()
        .nth_back(1)
        .map(|p| p.as_str())
        .unwrap_or(EMPTY);
    let grandparent = parts
        .iter()
        .nth_back(2)
        .map(|p| p.as_str())
        .unwrap_or(EMPTY);
    let great_grandparent = parts
        .iter()
        .nth_back(3)
        .map(|p| p.as_str())
        .unwrap_or(EMPTY);
    let parent_path = || path.parent().unwrap().to_path_buf();
    let grandparent_path = || parent_path().parent().unwrap().to_path_buf();
    let great_grandparent_path = || grandparent_path().parent().unwrap().to_path_buf();
    let great_great_grandparent_path = || great_grandparent_path().parent().unwrap().to_path_buf();
    let is_mise_dir = |d: &str| d == "mise" || d == ".mise";
    let is_config_filename = |f: &str| {
        f == "config.toml" || f == "config.local.toml" || regex!(r"config\..+\.toml").is_match(f)
    };
    let out = if parent == "conf.d" && is_mise_dir(grandparent) {
        if great_grandparent == ".config" {
            great_great_grandparent_path()
        } else {
            great_grandparent_path()
        }
    } else if is_mise_dir(parent) && is_config_filename(filename) {
        if grandparent == ".config" {
            great_grandparent_path()
        } else {
            grandparent_path()
        }
    } else if parent == ".config" {
        grandparent_path()
    } else {
        parent_path()
    };
    CONFIG_ROOT_CACHE.lock().unwrap().insert(path, out.clone());
    out
}

#[cfg(test)]
#[cfg(unix)]
mod tests {
    use super::*;

    #[test]
    fn test_config_root() {
        for p in &[
            "/foo/bar/.config/mise/conf.d/config.toml",
            "/foo/bar/.config/mise/conf.d/foo.toml",
            "/foo/bar/.config/mise/config.local.toml",
            "/foo/bar/.config/mise/config.toml",
            "/foo/bar/.config/mise.local.toml",
            "/foo/bar/.config/mise.toml",
            "/foo/bar/.mise.env.toml",
            "/foo/bar/.mise.local.toml",
            "/foo/bar/.mise.toml",
            "/foo/bar/.mise/conf.d/config.toml",
            "/foo/bar/.mise/config.local.toml",
            "/foo/bar/.mise/config.toml",
            "/foo/bar/.tool-versions",
            "/foo/bar/mise.env.toml",
            "/foo/bar/mise.local.toml",
            "/foo/bar/mise.toml",
            "/foo/bar/mise/config.local.toml",
            "/foo/bar/mise/config.toml",
            "/foo/bar/.config/mise/config.env.toml",
            "/foo/bar/.config/mise.env.toml",
            "/foo/bar/.mise/config.env.toml",
            "/foo/bar/.mise.env.toml",
        ] {
            println!("{p}");
            assert_eq!(config_root(Path::new(p)), PathBuf::from("/foo/bar"));
        }
    }
}
