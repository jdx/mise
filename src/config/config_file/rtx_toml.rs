use std::collections::HashMap;
use std::fmt::{Display, Formatter};
use std::fs;
use std::path::{Path, PathBuf};
use std::time::Duration;

use color_eyre::eyre::eyre;
use color_eyre::{Result, Section};
use log::LevelFilter;
use toml_edit::{table, value, Array, Item, Value};

use crate::config::config_file::{ConfigFile, ConfigFileType};
use crate::config::settings::SettingsBuilder;
use crate::config::{AliasMap, MissingRuntimeBehavior};
use crate::plugins::PluginName;
use crate::toolset::{ToolSource, ToolVersion, ToolVersionList, ToolVersionType, Toolset};

#[derive(Debug, Default)]
pub struct RtxToml {
    path: PathBuf,
    toolset: Toolset,
    env: HashMap<String, String>,
    settings: SettingsBuilder,
    alias: AliasMap,
    doc: toml_edit::Document,
    plugins: HashMap<String, String>,
}

#[macro_export]
macro_rules! parse_error {
    ($key:expr, $val:expr, $t:expr) => {{
        Err(eyre!(
            r#"expected value of "{}" to be a {}, got: {}"#,
            $key,
            $t,
            $val
        ))
    }};
}

#[allow(dead_code)] // TODO: remove
impl RtxToml {
    pub fn init(filename: &Path) -> Self {
        Self {
            path: filename.to_path_buf(),
            ..Default::default()
        }
    }

    pub fn from_file(filename: &Path) -> Result<Self> {
        trace!("parsing: {}", filename.display());
        let body = fs::read_to_string(filename).suggestion("ensure file exists and can be read")?;
        let mut rf = Self::from_str(body)?;
        rf.path = filename.to_path_buf();
        Ok(rf)
    }

    pub fn from_str(s: String) -> Result<Self> {
        Self {
            doc: s.parse().suggestion("ensure file is valid TOML")?,
            ..Default::default()
        }
        .parse()
    }

    fn parse(mut self) -> Result<Self> {
        for (k, v) in self.doc.iter() {
            match k {
                "env" => self.env = self.parse_hashmap(k, v)?,
                "alias" => self.alias = self.parse_alias(k, v)?,
                "tools" => self.toolset = self.parse_toolset(k, v)?,
                "settings" => self.settings = self.parse_settings(k, v)?,
                "plugins" => self.plugins = self.parse_hashmap(k, v)?,
                _ => warn!("unknown key: {}", k),
            }
        }
        Ok(self)
    }

    pub fn settings(&self) -> SettingsBuilder {
        SettingsBuilder::default()
    }

    fn parse_alias(&self, k: &str, v: &Item) -> Result<AliasMap> {
        match v.as_table_like() {
            Some(table) => {
                let mut aliases = AliasMap::new();
                for (plugin, table) in table.iter() {
                    let k = format!("{}.{}", k, plugin);
                    let plugin_aliases = aliases.entry(plugin.into()).or_default();
                    match table.as_table_like() {
                        Some(table) => {
                            for (from, to) in table.iter() {
                                match to.as_str() {
                                    Some(s) => {
                                        plugin_aliases.insert(from.into(), s.into());
                                    }
                                    _ => parse_error!(format!("{}.{}", k, from), v, "string")?,
                                }
                            }
                        }
                        _ => parse_error!(k, v, "table")?,
                    }
                }
                Ok(aliases)
            }
            _ => parse_error!(k, v, "table")?,
        }
    }

    fn parse_hashmap(&self, key: &str, v: &Item) -> Result<HashMap<String, String>> {
        match v.as_table_like() {
            Some(table) => {
                let mut env = HashMap::new();
                for (k, v) in table.iter() {
                    match v.as_str() {
                        Some(s) => {
                            env.insert(k.into(), s.into());
                        }
                        _ => parse_error!(key, v, "string")?,
                    }
                }
                Ok(env)
            }
            _ => parse_error!(key, v, "table"),
        }
    }

    fn parse_toolset(&self, key: &str, v: &Item) -> Result<Toolset> {
        let source = ToolSource::RtxToml(self.path.clone());
        let mut toolset = Toolset::new(source);

        match v.as_table_like() {
            Some(table) => {
                for (plugin, v) in table.iter() {
                    let k = format!("{}.{}", key, plugin);
                    let tvl = self.parse_tool_version_list(&k, v, &plugin.into())?;
                    toolset.versions.insert(plugin.into(), tvl);
                }
                Ok(toolset)
            }
            _ => parse_error!(key, v, "table")?,
        }
    }

    fn parse_tool_version_list(
        &self,
        key: &str,
        v: &Item,
        plugin_name: &PluginName,
    ) -> Result<ToolVersionList> {
        let source = ToolSource::RtxToml(self.path.clone());
        let mut tool_version_list = ToolVersionList::new(source);

        match v {
            Item::ArrayOfTables(v) => {
                for table in v.iter() {
                    for (tool, v) in table.iter() {
                        let k = format!("{}.{}", key, tool);
                        let tv = self.parse_tool_version(&k, v, plugin_name)?;
                        tool_version_list.versions.push(tv);
                    }
                }
            }
            v => match v.as_array() {
                Some(v) => {
                    for v in v.iter() {
                        let item = Item::Value(v.clone());
                        let tv = self.parse_tool_version(key, &item, plugin_name)?;
                        tool_version_list.versions.push(tv);
                    }
                }
                _ => {
                    tool_version_list
                        .versions
                        .push(self.parse_tool_version(key, v, plugin_name)?)
                }
            },
        }

        Ok(tool_version_list)
    }

    fn parse_tool_version(
        &self,
        key: &str,
        v: &Item,
        plugin_name: &PluginName,
    ) -> Result<ToolVersion> {
        let mut tv = ToolVersion::new(plugin_name.clone(), ToolVersionType::System);

        match v.as_table_like() {
            Some(table) => {
                if let Some(v) = table.get("version") {
                    match v {
                        Item::Value(v) => {
                            tv.r#type = self.parse_tool_version_value(key, v)?;
                        }
                        _ => parse_error!(format!("{}.version", key), v, "string")?,
                    }
                } else if let Some(path) = table.get("path") {
                    match path.as_str() {
                        Some(s) => {
                            tv.r#type = ToolVersionType::Path(s.into());
                        }
                        _ => parse_error!(format!("{}.path", key), v, "string")?,
                    }
                } else if let Some(prefix) = table.get("prefix") {
                    match prefix.as_str() {
                        Some(s) => {
                            tv.r#type = ToolVersionType::Prefix(s.into());
                        }
                        _ => parse_error!(format!("{}.prefix", key), v, "string")?,
                    }
                } else if let Some(r) = table.get("ref") {
                    match r.as_str() {
                        Some(s) => {
                            tv.r#type = ToolVersionType::Ref(s.into());
                        }
                        _ => parse_error!(format!("{}.ref", key), v, "string")?,
                    }
                } else {
                    parse_error!(key, v, "version, path, or prefix")?
                }
                for (k, v) in table.iter() {
                    if k == "version" || k == "path" || k == "prefix" || k == "ref" {
                        continue;
                    }
                    match v.as_str() {
                        Some(s) => {
                            tv.options.insert(k.into(), s.into());
                        }
                        _ => parse_error!(format!("{}.{}", key, k), v, "string")?,
                    }
                }
            }
            _ => match v {
                Item::Value(v) => {
                    tv.r#type = self.parse_tool_version_value(key, v)?;
                }
                _ => parse_error!(key, v, "value")?,
            },
        }

        Ok(tv)
    }

    fn parse_tool_version_value(&self, key: &str, v: &Value) -> Result<ToolVersionType> {
        match v.as_str() {
            Some(s) => match s.split_once(':') {
                Some(("prefix", v)) => Ok(ToolVersionType::Prefix(v.into())),
                Some(("path", v)) => Ok(ToolVersionType::Path(v.into())),
                Some(("ref", v)) => Ok(ToolVersionType::Ref(v.into())),
                Some((unknown, v)) => {
                    parse_error!(format!("{}.{}", key, unknown), v, "prefix, path, or ref")?
                }
                None => Ok(ToolVersionType::Version(s.into())),
            },
            _ => parse_error!(key, v, "string")?,
        }
    }

    fn parse_settings(&self, key: &str, v: &Item) -> Result<SettingsBuilder> {
        let mut settings = SettingsBuilder::default();

        match v.as_table_like() {
            Some(table) => {
                for (config_key, v) in table.iter() {
                    let k = format!("{}.{}", key, config_key);
                    match config_key.to_lowercase().as_str() {
                        "experimental" => settings.experimental = Some(self.parse_bool(&k, v)?),
                        "missing_runtime_behavior" => {
                            settings.missing_runtime_behavior =
                                Some(self.parse_missing_runtime_behavior(&k, v)?)
                        }
                        "legacy_version_file" => {
                            settings.legacy_version_file = Some(self.parse_bool(&k, v)?)
                        }
                        "always_keep_download" => {
                            settings.always_keep_download = Some(self.parse_bool(&k, v)?)
                        }
                        "plugin_autoupdate_last_check_duration" => {
                            settings.plugin_autoupdate_last_check_duration =
                                Some(self.parse_duration_minutes(&k, v)?)
                        }
                        "verbose" => settings.verbose = Some(self.parse_bool(&k, v)?),
                        "asdf_compat" => settings.asdf_compat = Some(self.parse_bool(&k, v)?),
                        "jobs" => settings.jobs = Some(self.parse_usize(&k, v)?),
                        "shorthands_file" => {
                            settings.shorthands_file = Some(self.parse_path(&k, v)?)
                        }
                        "disable_default_shorthands" => {
                            settings.disable_default_shorthands = Some(self.parse_bool(&k, v)?)
                        }
                        "log_level" => settings.log_level = Some(self.parse_log_level(&k, v)?),
                        "shims_dir" => settings.shims_dir = Some(self.parse_path(&k, v)?),
                        "raw" => settings.raw = Some(self.parse_bool(&k, v)?),
                        _ => parse_error!(k, v, "setting")?,
                    };
                }
            }
            None => parse_error!("settings", v, "table")?,
        }

        Ok(settings)
    }

    fn set_alias(&mut self, plugin: &str, from: &str, to: &str) {
        self.alias
            .entry(plugin.into())
            .or_default()
            .insert(from.into(), to.into());
        self.doc
            .entry("alias")
            .or_insert_with(table)
            .as_table_like_mut()
            .unwrap()
            .entry(plugin)
            .or_insert_with(table)
            .as_table_like_mut()
            .unwrap()
            .insert(from, value(to));
    }

    pub fn remove_alias(&mut self, plugin: &str, from: &str) -> Result<()> {
        if let Some(aliases) = self.doc.get_mut("alias").and_then(|v| v.as_table_mut()) {
            if let Some(plugin_aliases) = aliases.get_mut(plugin).and_then(|v| v.as_table_mut()) {
                self.alias.get_mut(plugin).unwrap().remove(from);
                plugin_aliases.remove(from);
                if plugin_aliases.is_empty() {
                    aliases.remove(plugin);
                    self.alias.remove(plugin);
                }
            }
            if aliases.is_empty() {
                self.doc.as_table_mut().remove("alias");
            }
        }
        Ok(())
    }

    fn parse_duration_minutes(&self, k: &str, v: &Item) -> Result<Duration> {
        match v.as_value() {
            Some(Value::String(s)) => Ok(humantime::parse_duration(s.value())?),
            Some(Value::Integer(i)) => Ok(Duration::from_secs(*i.value() as u64 * 60)),
            _ => Err(eyre!("expected {k} to be an integer, got: {v}")),
        }
    }

    fn parse_bool(&self, k: &str, v: &Item) -> Result<bool> {
        match v.as_value().map(|v| v.as_bool()) {
            Some(Some(v)) => Ok(v),
            _ => parse_error!(k, v, "boolean")?,
        }
    }

    fn parse_usize(&self, k: &str, v: &Item) -> Result<usize> {
        match v.as_value().map(|v| v.as_integer()) {
            Some(Some(v)) => Ok(v as usize),
            _ => Err(eyre!("expected {k} to be an integer, got: {v}")),
        }
    }

    fn parse_path(&self, k: &str, v: &Item) -> Result<PathBuf> {
        match v.as_value().map(|v| v.as_str()) {
            Some(Some(v)) => Ok(v.into()),
            _ => Err(eyre!("expected {k} to be a path, got: {v}")),
        }
    }

    fn parse_string(&self, k: &str, v: &Item) -> Result<String> {
        match v.as_value().map(|v| v.as_str()) {
            Some(Some(v)) => Ok(v.into()),
            _ => Err(eyre!("expected {k} to be a string, got: {v}")),
        }
    }

    fn parse_missing_runtime_behavior(&self, k: &str, v: &Item) -> Result<MissingRuntimeBehavior> {
        let v = self.parse_string("missing_runtime_behavior", v)?;
        match v.to_lowercase().as_str() {
            "warn" => Ok(MissingRuntimeBehavior::Warn),
            "ignore" => Ok(MissingRuntimeBehavior::Ignore),
            "prompt" => Ok(MissingRuntimeBehavior::Prompt),
            "autoinstall" => Ok(MissingRuntimeBehavior::AutoInstall),
            _ => Err(eyre!(
                "expected {k} to be one of: 'warn', 'ignore', 'prompt', 'autoinstall'. Got: {v}"
            )),
        }
    }

    fn parse_log_level(&self, k: &str, v: &Item) -> Result<LevelFilter> {
        let level = self.parse_string(k, v)?.parse()?;
        Ok(level)
    }
}

impl Display for RtxToml {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}", self.dump())
    }
}

impl ConfigFile for RtxToml {
    fn get_type(&self) -> ConfigFileType {
        ConfigFileType::RtxToml
    }

    fn get_path(&self) -> &Path {
        self.path.as_path()
    }

    fn plugins(&self) -> HashMap<PluginName, String> {
        self.plugins.clone()
    }

    fn env(&self) -> HashMap<String, String> {
        self.env.clone()
    }

    fn remove_plugin(&mut self, plugin: &PluginName) {
        self.toolset.versions.remove(plugin);
        if let Some(tools) = self.doc.get_mut("tools") {
            if let Some(tools) = tools.as_table_like_mut() {
                tools.remove(plugin);
                if tools.is_empty() {
                    self.doc.as_table_mut().remove("tools");
                }
            }
        }
    }

    fn replace_versions(&mut self, plugin_name: &PluginName, versions: &[String]) {
        if let Some(plugin) = self.toolset.versions.get_mut(plugin_name) {
            plugin.versions = versions
                .iter()
                .map(|v| ToolVersion::new(plugin_name.clone(), ToolVersionType::Version(v.clone())))
                .collect();
        }
        let tools = self
            .doc
            .entry("tools")
            .or_insert_with(table)
            .as_table_mut()
            .unwrap();

        if versions.len() == 1 {
            tools.insert(plugin_name, value(versions[0].clone()));
        } else {
            let mut arr = Array::new();
            for v in versions {
                arr.push(v);
            }
            tools.insert(plugin_name, Item::Value(Value::Array(arr)));
        }
    }

    fn save(&self) -> Result<()> {
        let contents = self.dump();
        Ok(fs::write(&self.path, contents)?)
    }

    fn dump(&self) -> String {
        self.doc.to_string()
    }

    fn to_toolset(&self) -> &Toolset {
        &self.toolset
    }

    fn settings(&self) -> SettingsBuilder {
        self.settings.clone()
    }

    fn aliases(&self) -> AliasMap {
        self.alias.clone()
    }
}

#[allow(dead_code)] // TODO: remove
const ENV_SUGGESTION: &str = r#"
[env]
FOO = "bar"
"#;

#[cfg(test)]
mod tests {
    use indoc::formatdoc;
    use insta::{assert_debug_snapshot, assert_display_snapshot};

    use crate::dirs;

    use super::*;

    #[test]
    fn test_fixture() {
        let cf = RtxToml::from_file(&dirs::HOME.join("fixtures/.rtx.toml")).unwrap();

        assert_debug_snapshot!(cf.env());
        assert_debug_snapshot!(cf.settings());
        assert_debug_snapshot!(cf.plugins());
        assert_debug_snapshot!(cf.toolset);
        assert_debug_snapshot!(cf.alias);

        assert_display_snapshot!(cf);
    }

    #[test]
    fn test_env() {
        let cf = RtxToml::from_str(formatdoc! {r#"
        [env]
        foo="bar"
        "#})
        .unwrap();

        assert_debug_snapshot!(cf.env(), @r###"
        {
            "foo": "bar",
        }
        "###);
        assert_display_snapshot!(cf);
    }

    #[test]
    fn test_set_alias() {
        let mut cf = RtxToml::from_str(formatdoc! {r#"
        [alias.nodejs]
        16 = "16.0.0"
        18 = "18.0.0"
        "#})
        .unwrap();

        cf.set_alias("nodejs", "18", "18.0.1");
        cf.set_alias("nodejs", "20", "20.0.0");
        cf.set_alias("python", "3.10", "3.10.0");

        assert_debug_snapshot!(cf.alias);
        assert_display_snapshot!(cf);
    }

    #[test]
    fn test_remove_alias() {
        let mut cf = RtxToml::from_str(formatdoc! {r#"
        [alias.nodejs]
        16 = "16.0.0"
        18 = "18.0.0"

        [alias.python]
        "3.10" = "3.10.0"
        "#})
        .unwrap();
        cf.remove_alias("nodejs", "16").unwrap();
        cf.remove_alias("python", "3.10").unwrap();

        assert_debug_snapshot!(cf.alias);
        assert_display_snapshot!(cf);
    }

    #[test]
    fn test_replace_versions() {
        let mut cf = RtxToml::from_str(formatdoc! {r#"
        [tools]
        nodejs = ["16.0.0", "18.0.0"]
        "#})
        .unwrap();
        cf.replace_versions(
            &PluginName::from("nodejs"),
            &["16.0.1".into(), "18.0.1".into()],
        );

        assert_debug_snapshot!(cf.toolset);
        assert_display_snapshot!(cf);
    }

    #[test]
    fn test_remove_plugin() {
        let mut cf = RtxToml::from_str(formatdoc! {r#"
        [tools]
        nodejs = ["16.0.0", "18.0.0"]
        "#})
        .unwrap();
        cf.remove_plugin(&PluginName::from("nodejs"));

        assert_debug_snapshot!(cf.toolset);
        assert_display_snapshot!(cf);
    }
}
