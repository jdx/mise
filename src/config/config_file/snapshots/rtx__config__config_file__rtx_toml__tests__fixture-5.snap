---
source: src/config/config_file/rtx_toml.rs
expression: cf.alias
---
{
    "nodejs": {
        "my_custom_node": "18",
    },
}
